bundles/org.openhab.binding.mikrotik/src/main/java/org/openhab/binding/mikrotik/internal/handler/MikrotikRouterosBridgeHandler.java

```diff
@@ -0,0 +1,271 @@
+    }
+
+    @Override
+    protected void updateStatus(ThingStatus status, ThingStatusDetail statusDetail, @Nullable String description) {
+        logger.trace("Attempt updating {} status to {}; detail = {}", getThing().getUID(), status, statusDetail);
+        if (status == ThingStatus.ONLINE
+                || (status == ThingStatus.OFFLINE && statusDetail == ThingStatusDetail.COMMUNICATION_ERROR)) {
+            scheduleRefreshJob();
+        } else if (status == ThingStatus.OFFLINE && statusDetail == ThingStatusDetail.CONFIGURATION_ERROR) {
+            cancelRefreshJob();
+        }
+        // update the status only if it's changed
+        ThingStatusInfo statusInfo = ThingStatusInfoBuilder.create(status, statusDetail).withDescription(description)
+                .build();
+        if (!statusInfo.equals(getThing().getStatusInfo())) {
+            super.updateStatus(status, statusDetail, description);
+        }
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("Disposing RouterOS bridge");
+        cancelRefreshJob();
+        if (routeros != null) {
+            routeros.stop();
+            routeros = null;
+        }
+    }
+
+    private void scheduleRefreshJob() {
+        synchronized (this) {
+            if (refreshJob == null) {
+                logger.debug("Scheduling refresh job every {}s", config.refresh);
+                refreshJob = scheduler.scheduleWithFixedDelay(this::scheduledRun, 0, config.refresh, TimeUnit.SECONDS);
+            }
+        }
+    }
+
+    private void cancelRefreshJob() {
+        synchronized (this) {
+            if (refreshJob != null) {
+                logger.debug("Cancelling refresh job");
+                refreshJob.cancel(true);
+                refreshJob = null;
+            }
+        }
+    }
+
+    private void scheduledRun() {
+        if (!routeros.isConnected()) {
+            // Perform connection
+            try {
+                logger.debug("Starting routeros model");
+                routeros.start();
+
+                @Nullable
+                RouterosRouterboardInfo rbInfo = routeros.getRouterboardInfo();
+                if (rbInfo != null) {
+                    Map<String, String> bridgeProps = editProperties();
+                    bridgeProps.put(PROPERTY_MODEL, rbInfo.getModel());
+                    bridgeProps.put(PROPERTY_FIRMWARE, rbInfo.getFirmware());
+                    bridgeProps.put(PROPERTY_SERIAL_NUMBER, rbInfo.getSerialNumber());
+                    updateProperties(bridgeProps);
+                } else {
+                    logger.warn("Failed to set RouterBOARD properties for bridge {}", getThing().getUID());
+                }
+                updateStatus(ThingStatus.ONLINE);
+            } catch (MikrotikApiException e) {
+                logger.warn("Error while logging in to RouterOS {}", getThing().getUID(), e);
+                logger.debug("RouterOS connection exception cause", e.getCause());
```

Comment:
If we come here also if the network fails it would seem unreasonable to log the stack trace. Aren't these messages already covered by `updateStatus()`?

Reply:
Not only network, but a variety of other unpredictable issues _may_ arise here...
Mikrotik API is kind of text/binary, and a user may try to connect to device with very old firmware, or some sort of new/beta firmware where password hashing algorithm could be different, or something like that.
Stacktrace will help to understand where exactly the API wrapper library has failed.

Reply:
OK, but please merge these two log messages into one to make clear to which error the cause belongs, in case concurrent errors arise.